### ISSUE TITLE
Ractive 0.8.0 Array wrapper crash

### ISSUE THREAD

**Issue by Jeff17Robbins** [ractivejs/ractive#2698]
## Description:
Array wrapper (`modifyArrays: true`) crash due to an attempt to teardown an already torndown array (in a somewhat complex use-case derived from a real application also using a Ractive-Ractive adaptor)

## Versions affected:
Ractive.js v0.8.1-build-4
Mon Oct 17 2016 14:48:43 GMT-0400 (EDT) - commit 24182db3673a1d349739e6aaf56a2fe4c9818c11

## Platforms affected:
Windows 10 Chrome Version 54.0.2840.59 m

## Reproduction:

https://jsfiddle.net/mdurpv9c/ (crashes in Chrome devtools debugger)

(but doesn't crash in Ractive 0.7.3) https://jsfiddle.net/ne4g3x2s/

stack trace of crash -- second teardown attempt crashes
teardown (ractive@0.8.1-build-4:10317)
adapt (ractive@0.8.1-build-4:2971)
mark$1 (ractive@0.8.1-build-4:3123)
update$2 (ractive@0.8.1-build-4:3580)
Ractive$update (ractive@0.8.1-build-4:3606)
(anonymous function) ((index):87)
each ((index):106)
onChange ((index):84)
notifySubscribers (ractive@0.8.1-build-4:1911)
fireEventAs (ractive@0.8.1-build-4:1876)
fireEvent (ractive@0.8.1-build-4:1853)
Ractive$fire (ractive@0.8.1-build-4:1928)
fire (ractive@0.8.1-build-4:734)
flushChanges (ractive@0.8.1-build-4:1290)
end (ractive@0.8.1-build-4:1196)
set (ractive@0.8.1-build-4:1374)
Ractive$set (ractive@0.8.1-build-4:16531)
window.onload ((index):170)

## **Code:**

HTML
```
<script src="https://unpkg.com/ractive@edge"></script>
<view-list></view-list>
```

`<view-list>` is a custom element, along with `<view-item>`, in the following JavaScript.  The problem seems to stem from having multiple (nested) references to the array.

The `<view-list>` element is populated via this template:
```
{{#each Store.someArray:idx}}<view-item idx="{{idx}}"></view-item>{{/each}}
```

This template stamps out a number of `<view-item>` elements, which are populated via this template:
```
{{Store.someArray[idx]}}
```

The Ractive-Ractive adaptor subscribes to the standard Ractive `change` event, and the problem seems to be that the `notifySubscribers(...)` loop delivering `change` events delivers the array to one subscriber, the old value's wrapper gets torndown, and then the next delivery ends up trying to teardown that already torndown array wrapper, causing a crash.

some of the JavaScript code (Ractive-Ractive adaptor code not shown, but at the jsfiddle [link](https://jsfiddle.net/mdurpv9c/))
```
// "store" -- a Ractive instance for holding data
var store = new Ractive({
  modifyArrays: true,
  adapt: ['Ractive']
});
store.set('someArray', ['a', 'b', 'c']);

// view-list -- a webcomponent containing a Ractive instance linked to the store via Ractive-Ractive
var elementDef1 = Object.create(HTMLElement.prototype);
elementDef1.createdCallback = function() {
  console.log('in createdCallback for view-list');
  this.style.display = 'block';
  this.ractive = new Ractive({
    onconfig: function() {
      this.set('Store', store);
    },
    el: this,
    template: '{{#each Store.someArray:idx}}<view-item idx="{{idx}}"></view-item>{{/each}}',
    modifyArrays: true,
    adapt: ['Ractive']
  });
  console.log(this.ractive.get());
  console.log(this.ractive.get('Store.someArray'));
};
elementDef1.detachedCallback = function() {
  console.log('in detachedCallback for view-list');
  this.ractive.teardown();
  this.ractive = null;
};
window.ViewList = document.registerElement('view-list', {
  prototype: elementDef1
});

// view-item -- a webcomponent containing a Ractive instance linked to the store via Ractive-Ractive
var elementDef2 = Object.create(HTMLElement.prototype);
elementDef2.createdCallback = function() {
  console.log('in createdCallback for view-item');
  this.style.display = 'block';
  this.ractive = new Ractive({
    onconfig: function() {
      this.set('Store', store);
    },
    el: this,
    template: '{{Store.someArray[idx]}}',
    modifyArrays: true,
    adapt: ['Ractive']
  });
  this.ractive.set('idx', this.getAttribute('idx'));
  console.log(this.ractive.get());
};
elementDef2.detachedCallback = function() {
  console.log('in detachedCallback for view-item');
  this.ractive.teardown();
  this.ractive = null;
};
window.ViewItem = document.registerElement('view-item', {
  prototype: elementDef2
});

// this causes the crash
store.set('someArray', []); 
```

**Comment by paulie4** [comment]
@Jeff17Robbins and I analyzed what's going on. `notifySubscribers()` is making a copy of `ractive._subs['change']` but three of the `on('change', cb)` callbacks are `off()`ed during the loop, so should not be called.

The reason this crash doesn't happen in 0.7.3 is because 0.7.3's r-0 Ractive's TransitionManager becomes the parent of r-1's TransitionManager, so the custom elements for r-2, r-3, and r-4 aren't removed until r-0's `flushChanges` and therefore the `off()` calls aren't happening until all the `notifySubscribers()` are called, whereas 0.8 is detaching them in r-1's `flushChanges`.

We think an easy way to fix this problem is to make the `onChange()` `ractive._subs` array have an indirect object to hold the reference to the callback, so if an off('change', cb) happens, then the indirect object's callback can be set to `noop` before it gets removed from `ractive._subs`. This will allow the `notifySubscribers()` loop to work as-is but the `off()`ed onChange would not be called.

**Comment by evs-chris** [comment]
Thanks for the analysis! I went the mildly hacky route and added an `off` flag to the subscriber callback when it is `off`ed and a check in `fireAsEvent` to make sure the callback hasn't been `off`ed. I _think_ that solves this issue, but I'm not positive due to the complexity of the issue.

I've also published 0.8.1-build-5, which includes this fix.

**Comment by evs-chris** [comment]
I'm going to close this, but feel free to re-open it if the issue pops back up in another way.
